shell/edit/provisioning.cattle.io.cluster/AddOnConfig.vue

```diff
@@ -0,0 +1,132 @@
+<script>
+import { Banner } from '@components/Banner';
+
+import Questions from '@shell/components/Questions';
+import YamlEditor from '@shell/components/YamlEditor';
+import { camelToTitle } from '@shell/utils/string';
+
+export default {
+  components: {
+    Banner,
+    Questions,
+    YamlEditor
+  },
+
+  props: {
+    mode: {
+      type:     String,
+      required: true,
+    },
+
+    value: {
+      type:     Object,
+      required: true,
+    },
+    isEdit: {
+      type:     Boolean,
+      required: true,
+    },
+
+    versionInfo: {
+      type:     Object,
+      required: true,
+    },
+
+    addonVersions: {
+      type:     Array,
+      required: false,
+      default:  null
+    },
+
+    addonsRev: {
+      type:     Number,
+      required: true,
+    },
+    userChartValuesTemp: {
+      type:     Object,
+      required: true,
+    },
+    initYamlEditor: {
+      type:     Function,
+      required: true,
+    }
+
+  },
+
+  computed: {
+    additionalManifest() {
+      return this.value.spec.rkeConfig.additionalManifest;
+    }
+  },
+
+  methods: {
+
+    labelForAddon(name) {
+      const fallback = `${ camelToTitle(name.replace(/^(rke|rke2|rancher)-/, '')) } Configuration`;
+
+      return this.$store.getters['i18n/withFallback'](`cluster.addonChart."${ name }"`, null, fallback);
+    },
+  },
+};
+</script>
+
+<template>
+  <div>
+    <Banner
+      v-if="isEdit"
+      color="warning"
+    >
+      {{ t('cluster.addOns.dependencyBanner') }}
+    </Banner>
+    <div
+      v-if="versionInfo && addonVersions.length"
+      :key="addonsRev"
+    >
+      <div
+        v-for="v in addonVersions"
+        :key="v._key"
+      >
+        <h3>{{ labelForAddon(v.name) }}</h3>
+        <Questions
+          v-if="versionInfo[v.name] && versionInfo[v.name].questions && v.name && userChartValuesTemp[v.name]"
+          v-model="userChartValuesTemp[v.name]"
+          :emit="true"
+          in-store="management"
+          :mode="mode"
+          :tabbed="false"
+          :source="versionInfo[v.name]"
+          :target-namespace="value.metadata.namespace"
+          @updated="$emit('update-questions', v.name)"
+        />
+        <YamlEditor
+          v-else
+          ref="yaml-values"
+          :value="initYamlEditor(v.name)"
+          :scrolling="true"
+          :as-object="true"
+          :editor-mode="mode === 'view' ? 'VIEW_CODE' : 'EDIT_CODE'"
+          :hide-preview-buttons="true"
+          @input="data => $emit('update-values', v.name, data)"
+        />
+        <div class="spacer" />
+      </div>
+    </div>
+
+    <div>
+      <h3>
+        {{ t('cluster.addOns.additionalManifest.title') }}
+        <i
+          v-clean-tooltip="t('cluster.addOns.additionalManifest.tooltip')"
+          class="icon icon-info"
+        />
+      </h3>
+      <YamlEditor
+        ref="yaml-additional"
```

Comment:
not sure this ref is actually used?

Reply:
I think it's used inside the actual YamlEditor component? When access through this.$refs

Reply:
this.$refs in the YamlEditor component wouldn't include this ref; only this.$refs in the AddonConfig component. I did eventually work out where this is used and why:

https://github.com/rancher/dashboard/blob/33381d4a310891142c829b53b39cea95dfcd3610/shell/edit/provisioning.cattle.io.cluster/rke2.vue#L1492-L1503 

If the YamlEditor is hidden with v-if (ie rendered but not displayed) the CodeMirror component inside it needs to be explicitly told to re-calculate its viewport when the YamlEditor is shown: that's what happens in the `refresh()` function quoted above.

With this YamlEditor now nested within the AddonConfig component the refreshYamls method in rke2.vue no longer calls refresh() on it when the tab is rendered. You can see how this causes problems by entering some info into 'Additional Manifest', saving the cluster, then going back to edit it: the yaml editor has a scroll bar and your data isn't in view until you click it:

https://github.com/rancher/dashboard/assets/42977925/499452dc-2aae-43da-aa24-d1e9d80aefde


You will need to make sure any instances of `YamlEditor` in tabs have the refresh() method run when the tab becomes active.